SleepKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsleepkotlin/MainActivity.kt

```diff
@@ -0,0 +1,620 @@
+ * granular sleep sessions, including light, deep, rem types of sleep, and also reading sleep
+ * sessions, including the use of aggregation to obtain summaries by day and for an entire period.
+ *
+ * For this example, seven nights of sleep data are written for a specific period of time, that
+ * falls within the period PERIOD_START_DATE_TIME to PERIOD_END_DATE_TIME. This bounding period is
+ * then used in querying the Fit API for sleep data in raw and aggregated forms.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions = FitnessOptions.builder()
+            .addDataType(DataType.TYPE_ACTIVITY_SEGMENT, FitnessOptions.ACCESS_READ)
+            .addDataType(DataType.TYPE_ACTIVITY_SEGMENT, FitnessOptions.ACCESS_WRITE)
+            .build()
+
+    private val runningQOrLater =
+            android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q
+
+    // Some responses from the Fit API are not filtered by activity type. This set is used as a
+    // filter to retain only those types relevant to sleep.
+    private val sleepActivities = setOf(
+            FitnessActivities.SLEEP,
+            FitnessActivities.SLEEP_LIGHT,
+            FitnessActivities.SLEEP_DEEP,
+            FitnessActivities.SLEEP_REM,
+            FitnessActivities.SLEEP_AWAKE
+    )
+
+    // Defines the start and end of the period of interest in this example.
+    private val periodStartMillis = millisFromRfc339DateString(PERIOD_START_DATE_TIME)
+    private val periodEndMillis = millisFromRfc339DateString(PERIOD_END_DATE_TIME)
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        // Put application specific code here.
+        setContentView(R.layout.activity_main)
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+
+        checkPermissionsAndRun(FitActionRequestCode.INSERT_SLEEP_SESSIONS)
+    }
+
+    /**
+     * Creates a list of data sets, one for each night of sleep for a given week, that is used in
+     * inserting data into Google Fit.
+     *
+     * A DSL is used here simply in order to make the definition of all seven days
+     * concise.
+     *
+     * @return A list of sleep data sets, one data set per night of sleep.
+     */
+    private fun createSleepDataSets(): List<DataSet> {
+        val dataSource = getSleepDataSource()
+
+        // The DSL allows data sets to be defined concisely by specifying just a start date/time
+        // then following with durations for each period of sleep type, to avoid the need for
+        // repetitive {@code DataSet} and {@code DataPoint} builders.
+        return dataSets(dataSource) {
+            dataSet("2020-01-20T23:00:00Z") {
```

Comment:
A lot of Kotlin developers don't know DSL yet and will probably try to read your DSL source code and get confused. 

With small samples, I worry about cognitive overload... so I try to focus on just the thing I want them to learn. This looks really cool, but you could an extension function with named arguments work just as well? (Then there is no DSL file and it would read just as well.):
(
                date = "2020-01-20T23:00:00Z",
                light = 60,
                deep = 60,
                light = 60,
                rem = 60,
                deep = 60,
                light = 120
)

Reply:
I definitely see where you're coming from, great point.

Named arguments would be great, but I can't repeat names within a parameter list can I?

The best I can see would be an extension off DataSource something like this:

```val dataSource = ...

val dataSet = dataSource.createDataSet(
    date = "2020-01-20T23:00:00Z",
    60 to light,
    60 to deep
    60 to light
    ...
)
```

using a `vararg` of `Pair` for each of the sleep components. I don't like how the pair syntax reads though, and wonder if that might be confusing? The DSL at least reads nicely when used, even if users don't want to dive into how it works. But yes, perhaps in SleepDsl.kt it could be confusing.

Alternatively, it could be done just using the builder boilerplate without any extensions or DSL but it would possibly be quite verbose?

Thoughts?

Reply:
As per discussion F2F, have implemented the `Pair`-based solution. Thanks!